src/main/groovy/com/morpheusdata/nutanix/prismelement/plugin/NutanixPrismElementPluginProvisionProvider.groovy

```diff
@@ -259,47 +282,430 @@ class NutanixPrismElementPluginProvisionProvider extends AbstractProvisionProvid
+			networkProxy.proxyHost = proxyConfiguration.proxyHost
+			networkProxy.proxyPassword = proxyConfiguration.proxyPassword
+			networkProxy.proxyUser = proxyConfiguration.proxyUser
+			networkProxy.proxyPort = (Integer) proxyConfiguration.proxyPort
+			networkProxy.proxyWorkstation = proxyConfiguration.proxyWorkstation
+		}
+		return networkProxy
+	}
+
+	private String getOrUploadImage(HttpApiClient client, Map authConfig, Cloud cloud, VirtualImage virtualImage, User createdBy, String datastoreId) {
+		def imageExternalId = null
+		def lockId = null
+		def location = null
+		def lockKey = "nutanix.imageupload.${cloud.regionCode}.${virtualImage?.id}".toString()
+		try {
+			//hold up to a 1 hour lock for image upload
+			lockId = context.acquireLock(lockKey, [timeout: 60l * 60l * 1000l, ttl: 60l * 60l * 1000l]).blockingGet()
+
+			// Check if it already exists
+			if (virtualImage) {
+				VirtualImageLocation virtualImageLocation
+				try {
+					virtualImageLocation = context.services.virtualImage.location.find(
+						new DataQuery()
+							.withFilter('virtualImage.id', virtualImage.id)
+							.withFilter('refType', 'ComputeZone')
+							.withFilter('refId', cloud.id)
+							.withFilter('imageRegion', cloud.regionCode)
+					)
+
+					imageExternalId = virtualImageLocation?.externalId
+				} catch (e) {
+					log.info("Error in findVirtualImageLocation.. could be not found ${e}", e)
+				}
+
+				// validate it exists in nutanix
+				if (imageExternalId) {
+					imageExternalId = NutanixPrismElementApiService.checkImageId(client, [zone: cloud], imageExternalId)
+				}
+			}
+
+			// Either we don't have a location or it's not uploaded to nutanix.
+			// Check if nutanix has the image already. If so, make a location for it
+			if (!imageExternalId && virtualImage.systemImage || virtualImage.userUploaded) {
+				def imageList = NutanixPrismElementApiService.listImages(client, authConfig)
+				if (imageList.success) {
+					def existingImage = imageList.data.find { it.status?.name == virtualImage.name }
+					if (existingImage) {
+						imageExternalId = existingImage.metadata?.uuid
+						VirtualImageLocation virtualImageLocation = new VirtualImageLocation([
+							virtualImage: virtualImage,
+							externalId  : imageExternalId,
+							imageRegion : cloud.regionCode,
+							code        : "nutanix.prism.image.${cloud.id}.${imageExternalId}",
+							internalId  : imageExternalId,
+							refId       : cloud.id,
+							refType     : 'ComputeZone'
+						])
+						location = context.async.virtualImage.location.create(virtualImageLocation, cloud).blockingGet()
+					}
+				}
+			}
+
+			// If nutanix didn't have it either, let's do the full upload
+			if (!imageExternalId) {
+				// Create the image
+				def cloudFiles = context.async.virtualImage.getVirtualImageFiles(virtualImage).blockingGet()
+				def imageFile = cloudFiles?.find { cloudFile -> cloudFile.name.toLowerCase().endsWith(".qcow2") }
```

Comment:
I know this is probably copy-paste but what if the image is an iso or raw or another type? This is where that bug I fixed manifested. IMO if qcow2 is the only allowed type you should bail here if you don't find it. Otherwise the create will fail later and it's harder to tell why.

Reply:
this matches existing behavior. going to leave as is.